### ISSUE TITLE
VS Code quick fixes aren't being applied

### ISSUE THREAD

**Issue by MariaSolOs** [rust-lang/rust-analyzer#13759]
**rust-analyzer version**: 0.4.1314-standalone (14492043d 2022-12-09)

**rustc version**: 1.65.0 (897e37553 2022-11-02)

Some quick fixes (such as importing a module) aren't being applied in VS Code. See below for an example:

https://user-images.githubusercontent.com/62502207/206883680-e16a047f-f5ee-401a-9a5c-41bda89557de.mov

**Comment by flodiebold** [comment]
You should be getting a choice between `std::fs`, `std::os::linux::fs` and `std::os::unix::fs` after clicking on "Import fs". Not sure why it doesn't show up for you :confused: Maybe related to the VSCode version?

**Comment by flodiebold** [comment]
This functionality goes through here:
https://github.com/rust-lang/rust-analyzer/blob/21e61bee8b74e93f14205f4a6c316db08f811e38/editors/code/src/client.ts#L252-L263
and should then call this:
https://github.com/rust-lang/rust-analyzer/blob/21e61bee8b74e93f14205f4a6c316db08f811e38/editors/code/src/commands.ts#L747-L756

**Comment by MariaSolOs** [comment]
> You should be getting a choice between `std::fs`, `std::os::linux::fs` and `std::os::unix::fs` after clicking on "Import fs". Not sure why it doesn't show up for you 😕 Maybe related to the VSCode version?

Yeah this was my experience up to this week, when this started happening 😿. Strangely enough, this doesn't happen with all quick fixes. For example, you can see how below the quick fix for adding the lifetime works, but the one for filling the `match` arms doesn't.

https://user-images.githubusercontent.com/62502207/206919708-4da577fd-cc51-421f-8ed1-685bd2c6473a.mov

**Comment by MariaSolOs** [comment]
The server trace also shows that the `codeAction` LSP request is being sent and processed correctly, something weird must be happening when applying the command itself.

```
[Trace - 9:55:41 AM] Sending request 'textDocument/codeAction - (95)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs"
    },
    "range": {
        "start": {
            "line": 7,
            "character": 13
        },
        "end": {
            "line": 7,
            "character": 13
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 7,
                        "character": 10
                    },
                    "end": {
                        "line": 7,
                        "character": 13
                    }
                },
                "message": "missing match arm: `A` and `B` not covered",
                "code": "missing-match-arm",
                "codeDescription": {
                    "href": "https://rust-analyzer.github.io/manual.html#missing-match-arm"
                },
                "severity": 1,
                "source": "rust-analyzer"
            },
            {
                "range": {
                    "start": {
                        "line": 7,
                        "character": 10
                    },
                    "end": {
                        "line": 7,
                        "character": 13
                    }
                },
                "message": "non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\nthe matched value is of type `Foo`",
                "data": {
                    "rendered": "error[E0004]: non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\n --> src/main.rs:8:11\n  |\n8 |     match foo {}\n  |           ^^^ patterns `Foo::A` and `Foo::B` not covered\n  |\nnote: `Foo` defined here\n --> src/main.rs:2:5\n  |\n1 | enum Foo {\n  |      ---\n2 |     A,\n  |     ^ not covered\n3 |     B,\n  |     ^ not covered\n  = note: the matched value is of type `Foo`\nhelp: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms\n  |\n8 ~     match foo {\n9 +         Foo::A | Foo::B => todo!(),\n10+     }\n  |\n\n"
                },
                "code": "Click for full compiler diagnostic",
                "codeDescription": {
                    "href": "rust-analyzer-diagnostics-view:/diagnostic%20message?1#file%3A%2F%2F%2FUsers%2Fmajosolano%2FDesktop%2Fshabling%2Fsrc%2Fmain.rs"
                },
                "severity": 1,
                "relatedInformation": [
                    {
                        "message": "`Foo` defined here",
                        "location": {
                            "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                            "range": {
                                "start": {
                                    "line": 1,
                                    "character": 4
                                },
                                "end": {
                                    "line": 1,
                                    "character": 5
                                }
                            }
                        }
                    },
                    {
                        "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`",
                        "location": {
                            "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 13
                                },
                                "end": {
                                    "line": 7,
                                    "character": 16
                                }
                            }
                        }
                    }
                ],
                "source": "rustc"
            },
            {
                "range": {
                    "start": {
                        "line": 7,
                        "character": 13
                    },
                    "end": {
                        "line": 7,
                        "character": 16
                    }
                },
                "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`",
                "code": "E0004",
                "codeDescription": {
                    "href": "https://doc.rust-lang.org/error-index.html#E0004"
                },
                "severity": 4,
                "relatedInformation": [
                    {
                        "message": "original diagnostic",
                        "location": {
                            "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            }
                        }
                    }
                ],
                "source": "rustc"
            }
        ],
        "triggerKind": 2
    }
}


[Trace - 9:55:41 AM] Received response 'textDocument/codeAction - (95)' in 2ms.
Result: [
    {
        "title": "Inline variable",
        "kind": "refactor.inline",
        "data": {
            "codeActionParams": {
                "textDocument": {
                    "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs"
                },
                "range": {
                    "start": {
                        "line": 7,
                        "character": 13
                    },
                    "end": {
                        "line": 7,
                        "character": 13
                    }
                },
                "context": {
                    "diagnostics": [
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            },
                            "severity": 1,
                            "code": "missing-match-arm",
                            "codeDescription": {
                                "href": "https://rust-analyzer.github.io/manual.html#missing-match-arm"
                            },
                            "source": "rust-analyzer",
                            "message": "missing match arm: `A` and `B` not covered"
                        },
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            },
                            "severity": 1,
                            "code": "Click for full compiler diagnostic",
                            "codeDescription": {
                                "href": "rust-analyzer-diagnostics-view:/diagnostic%20message?1#file%3A%2F%2F%2FUsers%2Fmajosolano%2FDesktop%2Fshabling%2Fsrc%2Fmain.rs"
                            },
                            "source": "rustc",
                            "message": "non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\nthe matched value is of type `Foo`",
                            "relatedInformation": [
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 1,
                                                "character": 4
                                            },
                                            "end": {
                                                "line": 1,
                                                "character": 5
                                            }
                                        }
                                    },
                                    "message": "`Foo` defined here"
                                },
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 7,
                                                "character": 13
                                            },
                                            "end": {
                                                "line": 7,
                                                "character": 16
                                            }
                                        }
                                    },
                                    "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`"
                                }
                            ],
                            "data": {
                                "rendered": "error[E0004]: non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\n --> src/main.rs:8:11\n  |\n8 |     match foo {}\n  |           ^^^ patterns `Foo::A` and `Foo::B` not covered\n  |\nnote: `Foo` defined here\n --> src/main.rs:2:5\n  |\n1 | enum Foo {\n  |      ---\n2 |     A,\n  |     ^ not covered\n3 |     B,\n  |     ^ not covered\n  = note: the matched value is of type `Foo`\nhelp: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms\n  |\n8 ~     match foo {\n9 +         Foo::A | Foo::B => todo!(),\n10+     }\n  |\n\n"
                            }
                        },
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 13
                                },
                                "end": {
                                    "line": 7,
                                    "character": 16
                                }
                            },
                            "severity": 4,
                            "code": "E0004",
                            "codeDescription": {
                                "href": "https://doc.rust-lang.org/error-index.html#E0004"
                            },
                            "source": "rustc",
                            "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`",
                            "relatedInformation": [
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 7,
                                                "character": 10
                                            },
                                            "end": {
                                                "line": 7,
                                                "character": 13
                                            }
                                        }
                                    },
                                    "message": "original diagnostic"
                                }
                            ]
                        }
                    ]
                }
            },
            "id": "inline_local_variable:RefactorInline:0"
        }
    },
    {
        "title": "Fill match arms",
        "kind": "quickfix",
        "data": {
            "codeActionParams": {
                "textDocument": {
                    "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs"
                },
                "range": {
                    "start": {
                        "line": 7,
                        "character": 13
                    },
                    "end": {
                        "line": 7,
                        "character": 13
                    }
                },
                "context": {
                    "diagnostics": [
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            },
                            "severity": 1,
                            "code": "missing-match-arm",
                            "codeDescription": {
                                "href": "https://rust-analyzer.github.io/manual.html#missing-match-arm"
                            },
                            "source": "rust-analyzer",
                            "message": "missing match arm: `A` and `B` not covered"
                        },
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            },
                            "severity": 1,
                            "code": "Click for full compiler diagnostic",
                            "codeDescription": {
                                "href": "rust-analyzer-diagnostics-view:/diagnostic%20message?1#file%3A%2F%2F%2FUsers%2Fmajosolano%2FDesktop%2Fshabling%2Fsrc%2Fmain.rs"
                            },
                            "source": "rustc",
                            "message": "non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\nthe matched value is of type `Foo`",
                            "relatedInformation": [
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 1,
                                                "character": 4
                                            },
                                            "end": {
                                                "line": 1,
                                                "character": 5
                                            }
                                        }
                                    },
                                    "message": "`Foo` defined here"
                                },
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 7,
                                                "character": 13
                                            },
                                            "end": {
                                                "line": 7,
                                                "character": 16
                                            }
                                        }
                                    },
                                    "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`"
                                }
                            ],
                            "data": {
                                "rendered": "error[E0004]: non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\n --> src/main.rs:8:11\n  |\n8 |     match foo {}\n  |           ^^^ patterns `Foo::A` and `Foo::B` not covered\n  |\nnote: `Foo` defined here\n --> src/main.rs:2:5\n  |\n1 | enum Foo {\n  |      ---\n2 |     A,\n  |     ^ not covered\n3 |     B,\n  |     ^ not covered\n  = note: the matched value is of type `Foo`\nhelp: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms\n  |\n8 ~     match foo {\n9 +         Foo::A | Foo::B => todo!(),\n10+     }\n  |\n\n"
                            }
                        },
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 13
                                },
                                "end": {
                                    "line": 7,
                                    "character": 16
                                }
                            },
                            "severity": 4,
                            "code": "E0004",
                            "codeDescription": {
                                "href": "https://doc.rust-lang.org/error-index.html#E0004"
                            },
                            "source": "rustc",
                            "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`",
                            "relatedInformation": [
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 7,
                                                "character": 10
                                            },
                                            "end": {
                                                "line": 7,
                                                "character": 13
                                            }
                                        }
                                    },
                                    "message": "original diagnostic"
                                }
                            ]
                        }
                    ]
                }
            },
            "id": "add_missing_match_arms:QuickFix:1"
        }
    },
    {
        "title": "Add this function's return type",
        "kind": "refactor.rewrite",
        "data": {
            "codeActionParams": {
                "textDocument": {
                    "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs"
                },
                "range": {
                    "start": {
                        "line": 7,
                        "character": 13
                    },
                    "end": {
                        "line": 7,
                        "character": 13
                    }
                },
                "context": {
                    "diagnostics": [
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            },
                            "severity": 1,
                            "code": "missing-match-arm",
                            "codeDescription": {
                                "href": "https://rust-analyzer.github.io/manual.html#missing-match-arm"
                            },
                            "source": "rust-analyzer",
                            "message": "missing match arm: `A` and `B` not covered"
                        },
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 10
                                },
                                "end": {
                                    "line": 7,
                                    "character": 13
                                }
                            },
                            "severity": 1,
                            "code": "Click for full compiler diagnostic",
                            "codeDescription": {
                                "href": "rust-analyzer-diagnostics-view:/diagnostic%20message?1#file%3A%2F%2F%2FUsers%2Fmajosolano%2FDesktop%2Fshabling%2Fsrc%2Fmain.rs"
                            },
                            "source": "rustc",
                            "message": "non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\nthe matched value is of type `Foo`",
                            "relatedInformation": [
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 1,
                                                "character": 4
                                            },
                                            "end": {
                                                "line": 1,
                                                "character": 5
                                            }
                                        }
                                    },
                                    "message": "`Foo` defined here"
                                },
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 7,
                                                "character": 13
                                            },
                                            "end": {
                                                "line": 7,
                                                "character": 16
                                            }
                                        }
                                    },
                                    "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`"
                                }
                            ],
                            "data": {
                                "rendered": "error[E0004]: non-exhaustive patterns: `Foo::A` and `Foo::B` not covered\n --> src/main.rs:8:11\n  |\n8 |     match foo {}\n  |           ^^^ patterns `Foo::A` and `Foo::B` not covered\n  |\nnote: `Foo` defined here\n --> src/main.rs:2:5\n  |\n1 | enum Foo {\n  |      ---\n2 |     A,\n  |     ^ not covered\n3 |     B,\n  |     ^ not covered\n  = note: the matched value is of type `Foo`\nhelp: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms\n  |\n8 ~     match foo {\n9 +         Foo::A | Foo::B => todo!(),\n10+     }\n  |\n\n"
                            }
                        },
                        {
                            "range": {
                                "start": {
                                    "line": 7,
                                    "character": 13
                                },
                                "end": {
                                    "line": 7,
                                    "character": 16
                                }
                            },
                            "severity": 4,
                            "code": "E0004",
                            "codeDescription": {
                                "href": "https://doc.rust-lang.org/error-index.html#E0004"
                            },
                            "source": "rustc",
                            "message": "ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms: ` {\n        Foo::A | Foo::B => todo!(),\n    }`",
                            "relatedInformation": [
                                {
                                    "location": {
                                        "uri": "file:///Users/majosolano/Desktop/shabling/src/main.rs",
                                        "range": {
                                            "start": {
                                                "line": 7,
                                                "character": 10
                                            },
                                            "end": {
                                                "line": 7,
                                                "character": 13
                                            }
                                        }
                                    },
                                    "message": "original diagnostic"
                                }
                            ]
                        }
                    ]
                }
            },
            "id": "add_return_type:RefactorRewrite:2"
        }
    }
]
```

**Comment by MariaSolOs** [comment]
Okay another update: I can only reproduce the bug in the latest VS Code Insiders (1.75.0-insider, commit f076d553210e9ea65e7bccbfa2f7a786c10fcab2), but not in the latest stable release (1.74.0).

**Comment by Mikastiv** [comment]
I have the same issue using the same code-insiders version as you so you might be right

**Comment by MariaSolOs** [comment]
@Mikastiv glad to know it isn't just me.

I created the respective issue in the VS Code repo btw: https://github.com/microsoft/vscode/issues/168912

**Comment by MariaSolOs** [comment]
Okay I think I have a fix: #13781

**Comment by MariaSolOs** [comment]
So my fix isn't actually working yet (`'codeAction/resolve'` seems to be invoked twice and the first request fails) but here's what I've learned so far:

- In the latest VS Code Insiders, code actions associated with a command (like [this](https://github.com/rust-lang/rust-analyzer/blob/4909d4fbdbf4e9c81c766edd5a0e38042154bc39/editors/code/src/client.ts#L223) and [this](https://github.com/rust-lang/rust-analyzer/blob/4909d4fbdbf4e9c81c766edd5a0e38042154bc39/editors/code/src/client.ts#L250), but not the ones taking [this route](https://github.com/rust-lang/rust-analyzer/blob/4909d4fbdbf4e9c81c766edd5a0e38042154bc39/editors/code/src/client.ts#L209)) aren't resolved. That is, the command returned from the code action request are never handled [here](https://github.com/rust-lang/rust-analyzer/blob/4909d4fbdbf4e9c81c766edd5a0e38042154bc39/editors/code/src/commands.ts#L796). This isn't happening in VS Code (non Insiders).
- Removing the dummy edits (ad done in #13781) seems to fix the issue and the commands are successfully applied, but the server seems to receive 2 code action resolve requests.
  - The first one has no data and doesn't seem to go through the client (that is, the code [here](https://github.com/rust-lang/rust-analyzer/blob/8f6d0cd1b00794bfb4e6047de98c2747807c6492/crates/rust-analyzer/src/handlers.rs#L1130) gets hit but [this line](https://github.com/rust-lang/rust-analyzer/blob/4909d4fbdbf4e9c81c766edd5a0e38042154bc39/editors/code/src/commands.ts#L798) doesn't). Because `data` is `None`, an error notification shows up.
  - The second request goes through the client handler and `data` is not `None`, and then the command is successfully applied.

I'm unsure of why/who is sending that first request. I imagine the dummy edit addressed the issue is some way, but I'm not sure. Looking at the [code action API](https://code.visualstudio.com/api/references/vscode-api#CodeAction), VS Code should invoke the command no matter if the dummy edit is there or not.

@mjbvz I noticed that you did some work around code actions lately (like https://github.com/microsoft/vscode/pull/163875). Do you think something might have changed recently that broke Rust?

**Comment by mjbvz** [comment]
@MariaSolOs You're using the LSP, right? If you have a trace @dbaeumer might know more about what is going on here

**Comment by MariaSolOs** [comment]
@mjbvz I am. Here are the traces.

- [code.txt](https://github.com/rust-lang/rust-analyzer/files/10249150/code.txt): From VS Code Version 1.74.1 (Universal), commit: 1ad8d514439d5077d2b0b7ee64d2ce82a9308e5a
- [insiders.txt](https://github.com/rust-lang/rust-analyzer/files/10249151/insiders.txt): From Insiders version: 1.75.0-insider (Universal), commit: edc432e920c3ec2c2af5e8a99b8e4b782633d298
[no-edits.txt](https://github.com/rust-lang/rust-analyzer/files/10249153/no-edits.txt): Same Insiders as above but with the changes in #13781

**Comment by MariaSolOs** [comment]
@mjbvz What I find a bit weird is the difference in behaviour despite that the extension hasn't updated its LSP dependencies ([these ones](https://github.com/microsoft/vscode-languageserver-node)). The only difference is the VS Code extension host.

**Comment by MariaSolOs** [comment]
Okay I now understand why the dummy edits are needed: [VS Code will resolve the code action request if the `edit` is missing](https://github.com/microsoft/vscode-languageserver-node/blob/1f111087dd3713f8720c6959afc5e2823c05b982/client/src/common/codeAction.ts#L46), and so this is why removing them results in [this](https://github.com/rust-lang/rust-analyzer/blob/8f6d0cd1b00794bfb4e6047de98c2747807c6492/crates/rust-analyzer/src/handlers.rs#L1123) getting hit twice.

So the problem is that VS Code is no longer applying the command after the edit, despite the docs saying that 

> A CodeAction must set either [edit](https://code.visualstudio.com/api/references/vscode-api#CodeAction.edit) and/or a [command](https://code.visualstudio.com/api/references/vscode-api#CodeAction.command). If both are supplied, the edit is applied first, then the command is executed.

**Comment by MariaSolOs** [comment]
_(Wow sorry for the spam lol I'm just invested in fixing this)_

#13781 fixes this problem. However, the root bug seems to live in VS Code: https://github.com/microsoft/vscode/issues/169507. I guess that we could wait for that one to be addressed, but note that the workaround of #13781 is quite small and will get code actions working today in all VS Code channels :)
